File: src/agent/skills/skill.py
"""Skill data model following Anthropic's Agent Skills convention.

Skills are directories containing a SKILL.md file with YAML frontmatter:

    .agent/skills/my-skill/SKILL.md
    ---
    name: my-skill
    description: Does something useful. Use when the user asks about X.
    user-invocable: true
    disable-model-invocation: false
    argument-hint: "[filename]"
    allowed-tools: Read, Grep, Glob
    ---

    Full instructions loaded when skill is invoked...
"""

from __future__ import annotations

from dataclasses import dataclass, field
from pathlib import Path


@dataclass
class Skill:
    """A skill following the Agent Skills specification."""
    name: str
    description: str = ""
    user_invocable: bool = True
    disable_model_invocation: bool = False
    argument_hint: str = ""
    allowed_tools: list[str] = field(default_factory=list)

    # Internal fields
    skill_dir: Path | None = None   # directory containing SKILL.md
    _body: str | None = None        # lazily loaded SKILL.md body (below frontmatter)

    def load_body(self) -> str:
        """Load the full SKILL.md body (instructions below frontmatter).

        Uses progressive loading: body is only read from disk on invocation.
        """
        if self._body is not None:
            return self._body

        if self.skill_dir is None:
            return ""

        skill_file = self.skill_dir / "SKILL.md"
        if not skill_file.exists():
            return ""

        content = skill_file.read_text()
        self._body = _extract_body(content)
        return self._body

    def render(self, arguments: str = "") -> str:
        """Render the skill body with argument substitution.

        Supports:
          $ARGUMENTS      - all arguments
          $ARGUMENTS[N]   - specific argument by index
          $0, $1, $2      - shorthand for $ARGUMENTS[N]
          !`command`       - dynamic context (shell command output)
        """
        body = self.load_body()
        if not body:
            return ""

        # Argument substitution
        if arguments:
            args_list = arguments.split()
            body = body.replace("$ARGUMENTS", arguments)
            for i, arg in enumerate(args_list):
                body = body.replace(f"$ARGUMENTS[{i}]", arg)
                body = body.replace(f"${i}", arg)
        elif "$ARGUMENTS" not in body and arguments:
            body += f"\nARGUMENTS: {arguments}"

        # Dynamic context: !`command` preprocessing
        body = _resolve_dynamic_context(body)

        return body


def _extract_body(content: str) -> str:
    """Extract the markdown body below YAML frontmatter."""
    if not content.startswith("---"):
        return content

    # Find closing ---
    end = content.find("---", 3)
    if end == -1:
        return content

    # Skip past the closing --- and any immediate newline
    body_start = end + 3
    if body_start < len(content) and content[body_start] == "\n":
        body_start += 1

    return content[body_start:].strip()


def _resolve_dynamic_context(body: str) -> str:
    """Replace !`command` patterns with shell command output."""
    import re
    import subprocess

    def run_command(match: re.Match) -> str:
        cmd = match.group(1)
        try:
            result = subprocess.run(
                cmd, shell=True, capture_output=True, text=True, timeout=10,
            )
            return result.stdout.strip() if result.returncode == 0 else f"(command failed: {cmd})"
        except subprocess.TimeoutExpired:
            return f"(command timed out: {cmd})"
        except Exception:
            return f"(command error: {cmd})"

    return re.sub(r"!`([^`]+)`", run_command, body)

File: src/skills/skill.ts
/** Skill data model following Anthropic's Agent Skills convention. */

import { execSync } from 'node:child_process';
import fs from 'node:fs';
import path from 'node:path';

export interface SkillInit {
  name: string;
  description?: string;
  userInvocable?: boolean;
  disableModelInvocation?: boolean;
  argumentHint?: string;
  allowedTools?: string[];
  skillDir?: string | null;
}

export class Skill {
  name: string;
  description: string;
  userInvocable: boolean;
  disableModelInvocation: boolean;
  argumentHint: string;
  allowedTools: string[];
  skillDir: string | null;
  private body: string | null = null;

  constructor(init: SkillInit) {
    this.name = init.name;
    this.description = init.description ?? '';
    this.userInvocable = init.userInvocable ?? true;
    this.disableModelInvocation = init.disableModelInvocation ?? false;
    this.argumentHint = init.argumentHint ?? '';
    this.allowedTools = init.allowedTools ?? [];
    this.skillDir = init.skillDir ?? null;
  }

  /** Load the full SKILL.md body (instructions below frontmatter). Lazy. */
  loadBody(): string {
    if (this.body !== null) return this.body;
    if (this.skillDir === null) return '';
    const skillFile = path.join(this.skillDir, 'SKILL.md');
    if (!fs.existsSync(skillFile)) return '';
    const content = fs.readFileSync(skillFile, 'utf-8');
    this.body = extractBody(content);
    return this.body;
  }

  /**
   * Render the skill body with argument substitution.
   *
   *   $ARGUMENTS      - all arguments
   *   $ARGUMENTS[N]   - specific argument by index
   *   $0, $1, $2      - shorthand for $ARGUMENTS[N]
   *   !`command`      - dynamic context (shell command output)
   */
  render(args = ''): string {
    let body = this.loadBody();
    if (!body) return '';

    if (args) {
      const argsList = args.split(/\s+/).filter((s) => s.length > 0);
      if (body.includes('$ARGUMENTS')) {
        body = body.split('$ARGUMENTS').join(args);
        argsList.forEach((arg, i) => {
          body = body.split(`$ARGUMENTS[${i}]`).join(arg);
          body = body.split(`$${i}`).join(arg);
        });
      } else {
        body += `\nARGUMENTS: ${args}`;
      }
    }

    body = resolveDynamicContext(body);
    return body;
  }
}

export function extractBody(content: string): string {
  if (!content.startsWith('---')) return content;
  const end = content.indexOf('---', 3);
  if (end === -1) return content;
  let bodyStart = end + 3;
  if (bodyStart < content.length && content[bodyStart] === '\n') bodyStart += 1;
  return content.slice(bodyStart).trim();
}

function resolveDynamicContext(body: string): string {
  return body.replace(/!`([^`]+)`/g, (_match, cmd: string) => {
    try {
      const out = execSync(cmd, { shell: '/bin/sh', timeout: 10_000, encoding: 'utf-8' });
      return out.trim();
    } catch (e: any) {
      if (e?.signal === 'SIGTERM' || e?.code === 'ETIMEDOUT') {
        return `(command timed out: ${cmd})`;
      }
      return `(command failed: ${cmd})`;
    }
  });
}
